File: data_db/src/main/java/br/org/restaurantepopular/data_db/repository/user/UserRepository.java
package br.org.restaurantepopular.data_db.repository.user;

import br.org.restaurantepopular.data_db.mapper.UserEntityMapper;
import br.org.restaurantepopular.entity.User;
import br.org.restaurantepopular.usecase.core.exception.NotFoundException;
import br.org.restaurantepopular.usecase.user.IUserRepository;

import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

public class UserRepository implements IUserRepository {

    private final UserDAO userDAO;

    public UserRepository(UserDAO userDAO) { this.userDAO = userDAO;}

    @Override
    public Optional<User> save (User user){
        return Optional.ofNullable(UserEntityMapper.toUser(
                userDAO.save(UserEntityMapper.toUserEntity(user))
        ));
    }

    @Override
    public void update(User user) throws NotFoundException {
        if(!userDAO.existsById(user.getId()))
            throw new NotFoundException(String.format("Not found a user with this %s ID",user.getId()));

        userDAO.save(UserEntityMapper.toUserEntity(user));
    }

    @Override
    public List<User> listUser() {
        return userDAO
                .findAll()
                .stream()
                .map(UserEntityMapper::toUser)
                .collect(Collectors.toList());
    }

    @Override
    public  Optional<User> findById(String id) throws NotFoundException {
        var userOptional = userDAO.findById(id).map(UserEntityMapper::toUser);
        if (userOptional.isEmpty())
            throw new NotFoundException(String.format("Not found a user with this %s ID", id));
        return userOptional;
    }

    @Override
    public void delete(String id){
        if (!userDAO.existsById(id))
            throw new NotFoundException(String.format("Not found a user with this %s ID", id));
        userDAO.deleteById(id);
    }






}

File: data_db/src/main/java/br/org/restaurantepopular/data_db/repository/user/UserRepository.kt
package br.org.restaurantepopular.data_db.repository.user

import br.org.restaurantepopular.usecase.core.exception.NotFoundException
import br.org.restaurantepopular.data_db.mapper.UserEntityMapper
import br.org.restaurantepopular.entity.User
import br.org.restaurantepopular.usecase.user.IUserRepository

class UserRepository(private val userDAO: UserDAO) : IUserRepository {

    override fun save(user: User): User {
        return UserEntityMapper.toUser(
            userDAO.save(UserEntityMapper.toUserEntity(user))
        )
    }

    @Throws(NotFoundException::class)
    override fun update(user: User) {
        if (!userDAO.existsById(user.id!!))
            throw NotFoundException(String.format("Not found a user with this %s ID", user.id))

        userDAO.save(UserEntityMapper.toUserEntity(user))
    }

    override fun listUser(): List<User> {
        return userDAO
            .findAll()
            .map(UserEntityMapper::toUser)
    }

    @Throws(NotFoundException::class)
    override fun findById(id: String): User {
        val user = userDAO.findById(id).map(UserEntityMapper::toUser)
        if (user.isEmpty)
            throw NotFoundException(String.format("Not found a user with this %s ID", id))
        return user.get()
    }

    override fun delete(id: String) {
        if (!userDAO.existsById(id))
            throw NotFoundException(String.format("Not found a user with this %s ID", id))
        userDAO.deleteById(id)
    }
}